### PR TITLE
remote-read: [2] Added chunked writer and reader.

### PR DESCRIPTION
**Pull request by bwplotka**
Second part of chained PRs to add streamed, chunked remote read as proposed here: https://docs.google.com/document/d/1JqrU3NjM9HoGLSTPYOvR217f5HBKBiJTqikEB9UiJL0/edit#

Part of https://github.com/prometheus/prometheus/issues/4517 and https://github.com/improbable-eng/thanos/issues/488

This is inspired by https://github.com/kythe/kythe/blob/master/kythe/go/platform/delimited/delimited.go but bit improved and tailored to our needs.

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: storage/remote/stream.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+// NewStreamReader constructs a StreamReader.
+func NewStreamReader(r io.Reader) *StreamReader {
+	return &StreamReader{b: bufio.NewReader(r)}
+}
+
+// Next returns the next length-delimited record from the input, or io.EOF if
+// there are no more records available. Returns io.ErrUnexpectedEOF if a short
+// record is found, with a length of n but fewer than n bytes of data.
+//
+// NOTE: The slice returned is valid only until a subsequent call to Next. It's a caller responsibility to copy the
```
Comment (brian-brazil) [inline storage/remote/stream.go]: the caller's

**Comment by bwplotka** [comment]
Fixed comments PTAL (:

File: storage/remote/stream.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	var buf [binary.MaxVarintLen64]byte
+	v := binary.PutUvarint(buf[:], uint64(len(b)))
+
+	if _, err := w.wrapped.Write(buf[:v]); err != nil {
+		return 0, err
+	}
+
+	n, err := w.wrapped.Write(b)
```
Comment (GiedriusS) [inline storage/remote/stream.go]: We could simplify this to a simple: `return w.wrapped.Write(b)`.

File: storage/remote/stream.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	if err != nil {
+		return n, err
+	}
+
+	return n, nil
```
Comment (tomwilkie) [inline storage/remote/stream.go]: Its hard to know without seeing the entire implementation, but I think you'll need a flush to ensure the data isn't buffered in the server.

**Comment by tomwilkie** [comment]
The HTTP server is probably going to be using HTTP chunked encoding under the hood here (hard to know for sure without the entire implementation), so is a custom framing format necessary?

**Comment by bwplotka** [comment]
> The HTTP server is probably going to be using HTTP chunked encoding under the hood here (hard to know for sure without the entire implementation), so is a custom framing format necessary?

Very good point, missed that, will explore this.

**Comment by bwplotka** [comment]
Closing until the alternative is checked (:

**Comment by bwplotka** [comment]
https://tools.ietf.org/html/rfc2616#section-3.6.1 :tada:

**Comment by bwplotka** [comment]
ok reopening, as after debugging a bit the HTTP can be chunked or whatever, but we have no control on how it chunks and IF it chunks so I think we need application layer chunk boundaries / delimiters like in this PR.

See: https://groups.google.com/d/msg/golang-nuts/y8CsrzluZmA/ioG5p07AwlcJ

**Comment by bwplotka** [comment]
That's why https://golang.org/pkg/encoding/json/#Decoder streaming decoder exists and there is nothing standard for protobuf

**Comment by bwplotka** [comment]
Also looked on gRPC internals and sure - they operate on TCP connection directly: https://github.com/grpc/grpc-go/blob/master/rpc_util.go#L592 https://github.com/grpc/grpc-go/blob/master/internal/transport/http2_server.go#L898 and put data frames as designed in HTTP/2.

If we choose to base on Chunked format, for HTTP golang client there are thousands edge cases on how it is chunked and for our protobuf parsing we need to know *exactly* number of bytes in each frame as it's the only info we have to tell what bytes to parse. And this info is actually deep in http `transferWire` struct: https://golang.org/src/net/http/transfer.go#L535

I am bit blocked on even simple MVP as not sure how to proceed as client response Body has to know exactly how much bytes to buffer in given buffer (as io.Reader) and we don't know that as we don't have access to underlying delimiter as specified in HTTP/1 rfc... unless we will operate directly on TCP conn instead HTTP Client which is overengineering IMO. ;p

**Comment by brian-brazil** [comment]
Sounds like we should do our own framing then.

**Comment by bwplotka** [comment]
Yea. Essentially the implementation for caller would be quite crazy for Golang, as Golang `http.Transport` does guarantee the `Read` will return chunk by chunk (although usually that's the case).   Short tests confirms it. With flush I can see the internal `ChunkedReader` is reading chunk by chunk (numbers are lenght of byte writes), but `response.Body.Read` is not giving me chunk by chunk strictly:
```
write  86
flush
write  87
flush
write  87
flush
(inside chunkedReader) READING  86
response.Body.Read 86
chunked_series:<labels:<name:"__name__" value:"test_metric1" > labels:<name:"b" value:"c" > labels:<name:"baz" value:"qux" > labels:<name:"d" value:"e" > labels:<name:"foo" value:"bar" > chunks:<type:XOR data:"\000\001\000?\360\000\000\000\000\000\000\000" > > 
(inside chunkedReader)  READING ..  87
(inside chunkedReader)  READING ..  87
response.Body.Read 174 & err EOF
``` 
Last one violates our protocol as from 174 bytes I have no idea as a caller if it's one protobuf or many.

Also worth to not that for this test I couldn't use `httptest.Recorder` in tests as it does not have any notion of internal chunking ;p 

Full attempt code is here: https://github.com/prometheus/prometheus/compare/master...bwplotka:rr-attempt-chunked-framing?expand=1#diff-c32caf482fb76fa695ceffe3455c6d9aR1059  I would not spend more time on it though.

Alternatively we could dial natively and setup TCP connection and do things directly on connection using https://golang.google.cn/src/net/http/internal/chunked.go but it is overkill and IMO it might be similarly hard for users of our remote read in different languages. 

I would stick to application delimition. Obviously we will rely on HTTP chunking for actual streaming, but the actual chunking is very close to how we "Flush()"  response on server side (but no strictly the same), so I think it's good enough. Varint should be simple for callers to expect as well, and the overhead should be small if we keep frames not too small.

File: storage/remote/stream.go (hunk 3 above)
Comment (bwplotka) [reply storage/remote/stream.go]: Yes. I planned to do it on caller side, but happy to ensure this here. Let's see what is more clear.

**Comment by bwplotka** [comment]
PTAL @tomwilkie @brian-brazil @SuperQ addressed comments. Moved to `ChunkedWriter/Reader` and added flushing.

**Comment by bwplotka** [comment]
Closing -> the squashed commits are now part of this PR: https://github.com/prometheus/prometheus/pull/5703

**Pull request closed without merging**